src/wasm_udf/wasm.rs

```diff
@@ -11,22 +15,518 @@ use datafusion::{
+        // deallocate both input and output buffers
+        self.dealloc
+            .call(&mut self.store, (udf_input_ptr, input_size))?;
+        self.dealloc
+            .call(&mut self.store, (udf_output_ptr, output_size))?;
+        Ok(output)
+    }
+}
+
+fn get_arrow_value<T>(
+    args: &[ArrayRef],
+    row_ix: usize,
+    col_ix: usize,
+) -> Result<T::Native>
+where
+    T: ArrowPrimitiveType,
+{
+    match args
+        .get(col_ix)
+        .unwrap()
+        .as_any()
+        .downcast_ref::<PrimitiveArray<T>>()
+    {
+        Some(arr) => Ok(arr.value(row_ix)),
+        None => Err(DataFusionError::Internal(format!(
+            "Error casting column {:?} to array of primitive values",
+            col_ix
         ))),
     }
 }
 
+fn messagepack_encode_input_value(
+    value_type: &CreateFunctionDataType,
+    args: &[ArrayRef],
+    row_ix: usize,
+    col_ix: usize,
+) -> Result<Value> {
+    match value_type {
+        CreateFunctionDataType::SMALLINT => {
+            get_arrow_value::<arrow::datatypes::Int16Type>(args, row_ix, col_ix)
+                .map(Value::from)
+        }
+        CreateFunctionDataType::I32 | CreateFunctionDataType::INT => {
+            get_arrow_value::<arrow::datatypes::Int32Type>(args, row_ix, col_ix)
+                .map(Value::from)
+        }
+        CreateFunctionDataType::I64 | CreateFunctionDataType::BIGINT => {
+            get_arrow_value::<arrow::datatypes::Int64Type>(args, row_ix, col_ix)
+                .map(Value::from)
+        }
+        CreateFunctionDataType::F32
+        | CreateFunctionDataType::FLOAT
+        | CreateFunctionDataType::REAL => {
+            get_arrow_value::<arrow::datatypes::Float32Type>(args, row_ix, col_ix)
+                .map(|val| Value::from(val.to_bits()))
+        }
+        CreateFunctionDataType::F64 | CreateFunctionDataType::DOUBLE => {
+            get_arrow_value::<arrow::datatypes::Float64Type>(args, row_ix, col_ix)
+                .map(|val| Value::from(val.to_bits()))
+        }
+        CreateFunctionDataType::TEXT
+        | CreateFunctionDataType::CHAR
+        | CreateFunctionDataType::VARCHAR => match args
+            .get(col_ix)
+            .unwrap()
+            .as_any()
+            .downcast_ref::<arrow::array::StringArray>(
+        ) {
+            Some(arr) => Ok(Value::from(arr.value(row_ix))),
+            None => Err(DataFusionError::Internal(format!(
+                "Error casting column {:?} to string array",
+                col_ix
+            ))),
+        },
+        CreateFunctionDataType::BOOLEAN => match args
+            .get(col_ix)
+            .unwrap()
+            .as_any()
+            .downcast_ref::<arrow::array::BooleanArray>(
+        ) {
+            Some(arr) => Ok(Value::from(arr.value(row_ix))),
+            None => Err(DataFusionError::Internal(format!(
+                "Error casting column {:?} to boolean array",
+                col_ix
+            ))),
+        },
```

Comment:
Does the `get_arrow_value` function not work in these cases?

Reply:
I see, looks like it's dual with encoding back from messagepack to arrow (where StringArray/BooleanArray aren't primitive arrays)